### PR TITLE
rest: Fix invalid query validation

### LINKED ISSUE #240: resource query sometimes raise 500 instead of 400 on invalid queries
I found some queries that returns 500 instead of 400:

```
ne:
  project_id: ["foobar"]
```

or

`not: [{"foobar": "barfoo"}]`

### PR DESCRIPTION
**Pull request by sileht**
Some operators accept a list or a dict at rest layer, when it should
accept only an attribute value, like bool, str, None.

For Postgresql, that result to an sql error, turned in HTTP 500.
For Mysql, the request succeed with some magic casting.

Closes #240

### CODE REVIEW

**Comment by sileht** [comment]
Pull-request updated, HEAD is now fac2e163ba9c997f6910587b4310d41678f7e7c7

**Comment by sileht** [comment]
Pull-request updated, HEAD is now db5530b54b5ed5910e0768e5813f81296f80b601

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 7f423a4f5172496b02068069b1e392e248696d77

File: gnocchi/rest/__init__.py [hunk 1]
```diff
@@ -1202,21 +1208,26 @@ def _ResourceSearchSchema():
...
+                ),
                 voluptuous.Any(
                     u"and", u"∨",
                     u"or", u"∧",
-                    u"not",
                 ): voluptuous.All(
                     [ResourceSearchSchema], voluptuous.Length(min=1)
-                )
+                ),
+                u"not": ResourceSearchSchema,
```
Comment (jd) [inline gnocchi/rest/__init__.py]: It'd be great to have a test for his change. Unless it's the `nor` one?

File: gnocchi/tests/functional/gabbits/search.yaml [hunk 2]
```diff
@@ -21,12 +21,39 @@ tests:
...
+        ne:
+          project_id:
+            - foobar
+      status: 400
+
+    - name: search invalid nor value
+      desc: uninary operator must follow by dict, not list
+      POST: /v1/search/resource/generic
+      data:
+        nor:
```
Comment (jd) [inline gnocchi/tests/functional/gabbits/search.yaml]: nor?

File: gnocchi/tests/functional/gabbits/search.yaml [hunk 3]
```diff
@@ -21,12 +21,39 @@ tests:
...
+    - name: search invalid ne value
+      desc: attribute value for binary operator must not be dict or list
+      POST: /v1/search/resource/generic
+      data:
+        ne:
+          project_id:
+            - foobar
+      status: 400
+
+    - name: search invalid nor value
```
Comment (jd) [inline gnocchi/tests/functional/gabbits/search.yaml]: nor?

File: gnocchi/tests/functional/gabbits/search.yaml (hunk 2 above)
Comment (sileht) [inline gnocchi/tests/functional/gabbits/search.yaml]: Fixed, I have also added some expected response string, to ensure I test the expected query.

File: gnocchi/tests/functional/gabbits/search.yaml (hunk 3 above)
Comment (sileht) [inline gnocchi/tests/functional/gabbits/search.yaml]: it's not, fixed.

File: gnocchi/rest/__init__.py (hunk 1 above)
Comment (sileht) [inline gnocchi/rest/__init__.py]: it's the nor one

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 6c151b8ab117864bd891f27c875bc174d49c4521

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 84f2fb8deee6e8ef4874b029530961d64dcec3e0

File: gnocchi/tests/functional/gabbits/search.yaml [hunk 4]
```diff
@@ -55,15 +92,22 @@ tests:
       response_json_paths:
         $.`len`: 2
 
-    - name: search like created_by_project_id
+    - name: search eq created_by_project_id
```
Comment (chungg) [inline gnocchi/tests/functional/gabbits/search.yaml]: we're actually missing a `like` test now.

**Pull request merged**